sycl/plugins/esimd_cpu/pi_esimd_cpu.cpp

```diff
@@ -593,367 +586,529 @@ pi_result piContextCreate(const pi_context_properties *Properties,
                             void *HostPtr, pi_mem *RetMem,
                             const pi_mem_properties *properties) {
-  (void)Context;
-  (void)Flags;
-  (void)Size;
-  (void)HostPtr;
-  (void)RetMem;
-  (void)properties;
-  DIE_NO_IMPLEMENTATION;
+  if ((Flags & PI_MEM_FLAGS_ACCESS_RW) == 0) {
+    return PI_INVALID_OPERATION;
+  }
+  if (Context == nullptr) {
+    return PI_INVALID_CONTEXT;
+  }
+  if (RetMem == nullptr) {
+    return PI_INVALID_VALUE;
+  }
+
+  cm_support::CmBuffer *CmBuf = nullptr;
+  cm_support::SurfaceIndex *CmIndex;
+
+  int Status = Context->Device->CmDevicePtr->CreateBuffer(
+      static_cast<unsigned int>(Size), CmBuf);
+
+  if (Status != cm_support::CM_SUCCESS) {
+    return PI_OUT_OF_HOST_MEMORY;
+  }
+
+  Status = CmBuf->GetIndex(CmIndex);
+
+  if ((Flags & PI_MEM_FLAGS_HOST_PTR_USE) != 0 ||
+      (Flags & PI_MEM_FLAGS_HOST_PTR_COPY) != 0) {
+    Status =
+        CmBuf->WriteSurface(reinterpret_cast<const unsigned char *>(HostPtr),
+                            nullptr, static_cast<unsigned int>(Size));
+  }
+
+  auto HostPtrOrNull =
+      (Flags & PI_MEM_FLAGS_HOST_PTR_USE) ? pi_cast<char *>(HostPtr) : nullptr;
+
+  try {
+    *RetMem =
+        new _pi_buffer(Context, HostPtrOrNull, CmBuf,
+                       /* integer buffer index */ CmIndex->get_data(), Size);
+  } catch (const std::bad_alloc &) {
+    return PI_OUT_OF_HOST_MEMORY;
+  } catch (...) {
+    return PI_ERROR_UNKNOWN;
+  }
+
   return PI_SUCCESS;
 }
 
 pi_result piMemGetInfo(pi_mem, cl_mem_info, size_t, void *, size_t *) {
   DIE_NO_IMPLEMENTATION;
-  return PI_SUCCESS;
 }
 
-pi_result piMemRetain(pi_mem) {
-  DIE_NO_IMPLEMENTATION;
+pi_result piMemRetain(pi_mem Mem) {
+  if (Mem == nullptr) {
+    return PI_INVALID_MEM_OBJECT;
+  }
+  ++(Mem->RefCount);
   return PI_SUCCESS;
 }
 
 pi_result piMemRelease(pi_mem Mem) {
-  (void)Mem;
-  DIE_NO_IMPLEMENTATION;
+  if (Mem == nullptr) {
+    return PI_INVALID_MEM_OBJECT;
+  }
+
+  Mem->RefCount -= 1;
+
+  if (Mem->RefCount != 0) {
+    return PI_SUCCESS;
+  }
+
+  if (Mem->getMemType() == PI_MEM_TYPE_BUFFER) {
+    _pi_buffer *PiBuf = static_cast<_pi_buffer *>(Mem);
+    int Result =
+        Mem->Context->Device->CmDevicePtr->DestroySurface(PiBuf->CmBufferPtr);
+
+    if (Result != cm_support::CM_SUCCESS) {
+      return PI_INVALID_MEM_OBJECT;
+    }
+  } else if (Mem->getMemType() == PI_MEM_TYPE_IMAGE2D) {
+    _pi_image *PiImg = static_cast<_pi_image *>(Mem);
+    int Result =
+        Mem->Context->Device->CmDevicePtr->DestroySurface(PiImg->CmSurfacePtr);
+
+    if (Result != cm_support::CM_SUCCESS) {
+      return PI_INVALID_MEM_OBJECT;
+    }
+  } else if (Mem->getMemType() == PI_MEM_TYPE_IMAGE2D) {
+    _pi_image *PiImg = static_cast<_pi_image *>(Mem);
+    int Result =
+        Mem->Context->Device->CmDevicePtr->DestroySurface(PiImg->CmSurfacePtr);
+    if (Result != cm_support::CM_SUCCESS) {
+      return PI_INVALID_MEM_OBJECT;
+    }
```

Comment:
Is there any different in these two branches? Can we unify them?

Reply:
Will remove.